### PR TITLE
msi: migration to c:\opt\fluent

### PR DESCRIPTION
**Pull request by kenhys**
msi: migrate to c:\opt\fluent

Before:

* Configuration c:\opt\td-agent\etc\td-agent\td-agent.conf
* Log files: c:\opt\td-agent

After:

* Move configuration to c:\opt\fluent\etc\fluent\td-agent.conf
* Create symbolic link to c:\opt\fluent\etc\fluent from c:\opt\td-agent\etc\td-agent.
* Overwrite c:\opt\fluent\etc\fluentd.conf with td-agent.conf (when upgrade)
* Migrate to FLUENT_PACKAGE_TOPDIR environment variable.
* Keep old logs as is:

### CODE REVIEW

**Comment by kenhys** [comment]
[debug-msi.log](https://github.com/fluent/fluent-package-builder/files/11785407/debug-msi.log)

InstallExecuteSequence doesn't work as expected :thinking:

**Comment by kenhys** [comment]
newly install works as expected, not for upgrade.
[pr494-new-install.log](https://github.com/fluent/fluent-package-builder/files/11785719/pr494-new-install.log)
[pr494-upgrade.log](https://github.com/fluent/fluent-package-builder/files/11785721/pr494-upgrade.log)

**Comment by kenhys** [comment]
Unexpectedly action was skipped 

```
MSI (s) (50:1C) [16:52:42:112]: Note: 1: 2265 2:  3: -2147287035 
MSI (s) (50:1C) [16:52:42:112]: Server not locked: locking for product {4FEF158E-E67C-4D05-B7E2-55DA1B33AF9F}
Action ended 16:52:42: InstallInitialize. Return value 1.
MSI (s) (50:1C) [16:52:42:125]: Skipping action: SetDeleteCompatTdAgentBat (condition is false)
MSI (s) (50:1C) [16:52:42:125]: Skipping action: DeleteCompatTdAgentBat (condition is false)
MSI (s) (50:1C) [16:52:42:125]: Skipping action: SetDeleteCompatTdAgentGemBat (condition is false)
MSI (s) (50:1C) [16:52:42:125]: Skipping action: DeleteCompatTdAgentGemBat (condition is false)
MSI (s) (50:1C) [16:52:42:125]: Doing action: ProcessComponents
MSI (s) (50:1C) [16:52:42:125]: Note: 1: 2205 2:  3: ActionText 
Action 16:52:42: ProcessComponents. Updating component registration
Action start 16:52:42: ProcessComponents.
```

**Comment by kenhys** [comment]
PostInstall action is called.

```
MSI (s) (50:1C) [16:53:06:342]: Executing op: ActionStart(Name=PostInstall,,)
Action 16:53:06: PostInstall. 
MSI (s) (50:1C) [16:53:06:344]: Executing op: CustomActionSchedule(Action=PostInstall,ActionType=3073,Source=BinaryData,Target=WixQuietExec64,CustomActionData="c:\opt\fluent\bin\fluent-package-post-install.bat")
MSI (s) (50:B4) [16:53:06:352]: Invoking remote custom action. DLL: C:\WINDOWS\Installer\MSICEAC.tmp, Entrypoint: WixQuietExec64
WixQuietExec64:  Restoring gems to pristine condition...
WixQuietExec64:  Skipped error_highlight-0.5.1, it is a default gem
```

**Comment by kenhys** [comment]
postinstall's `if exist "%~dp0..\..\td-agent" (` condition may not be evaluated expectedly :thinking:

**Comment by kenhys** [comment]
check error log:
[upgrade.take1.rollback.zip](https://github.com/fluent/fluent-package-builder/files/11796121/upgrade.take1.rollback.zip)

**Comment by kenhys** [comment]
> check error log: [upgrade.take1.rollback.zip](https://github.com/fluent/fluent-package-builder/files/11796121/upgrade.take1.rollback.zip)

Should align not expected WixQuietExec64 execution order.

**Comment by kenhys** [comment]
[upgrade-echo-on.u8.log](https://github.com/fluent/fluent-package-builder/files/11815470/upgrade-echo-on.u8.log)
tracked via @echo on, but it seems that the second `if exist ...` sequence is not executed.

**Comment by kenhys** [comment]
It is out of scope, I've found that post-install was always called because of a mismatch...

```
WixQuietExec64:  C:\Windows\SysWOW64>if not "c:\opt\fluent\bin\" == "C:\opt\fluent\bin\" ("c:\opt\fluent\bin\gem" pristine --only-executables --all ) 
WixQuietExec64:  Restoring gems to pristine condition...
```

**Comment by kenhys** [comment]
When executing extracted bat file content to simplified msi, it will be kicked. (even though weird mojibake errror)

**Comment by kenhys** [comment]
with 7fb0059, WixQuietExec64 failure was fixed, but it works still unexpectdly, checking why...

**Comment by kenhys** [comment]
c:\opt\td-agent was linked to c:\opt\fluent.

image

**Comment by kenhys** [comment]
image

c:\opt\fluent\var\log\td-agent was linked to c:\opt\fluent\var\log\fluent.

**Comment by daipom** [comment]
Thanks for this fix!
I will review this early next week.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:30]: Is this typo?

```suggestion
  if %ERRORLEVEL% == 1 (
```

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:33]: I think we can use `net stop` to stop the service synchronously.
(`net` command would be very famous and commonly used command, but I can't find the official document of this command...)

```suggestion
    net stop fluentdwinsvc
```

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:30]: Shouldn't this be the first process?
(right after the `if exist "%~dp0..\..\td-agent"`)

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:39]: We may need to consider the possibility of renaming it instead of deleting it.

**Review by daipom — COMMENTED** [review]
I commented on some points of concern.
I'm going to look at it in more detail, checking the actual behavior.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:30]: This is intentional, but I forgot to enable delayedexpansion.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:33]: Thank!

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:30]: Oh! delayedexpansion! I see!

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:45]: We also need the link `/opt/fluent/etc/td-agent` -> `/opt/fluent/etc/fluent` so that we can access the old conf file by `/opt/td-agent/etc/td-agent/td-agent.conf`.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:19]: > Move log files to c:\opt\fluent\var\log\fluent.

I think this is the logic.
This logic certainly moves `var\log\` files, but they are actually the buffer files only.
Currently, log files are directly under `C:\opt\td-agent\`, and they are lost!
We should consider the log files migration.

* `C:\opt\td-agent\td-agent-supervisor-0.log`
* `C:\opt\td-agent\td-agent-0.log`

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat]: It seems better to change. thanks.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat]: Good catch.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:61]: Thanks, right thing to do so.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:27]: Thanks. I'll change it.

**Review by daipom — COMMENTED** [review]
> Move configuration files to c:\opt\fluent\etc\fluent\td-agent.conf

I have a question about this specification.
With only this, Fluentd will start with the default setting `c:\opt\fluent\etc\fluent\fluentd.conf` after the update.

In the deb/rpm package, the path to the configuration file is changed by an environment variable.
That is difficult for Windows, so I guess that we put that on hold in this PR, right?

I think we should consider overwriting the new default config file with the old file. 
I think we could consider it in common with msi, deb and rpm later.

**Comment by kenhys** [comment]
also updated commit log.

**Comment by daipom** [comment]
If the old td-agent service is running, this pop-up appears.
(Maybe it's been that way before. I'll check it.)

2023-06-27_13h33_46

**Comment by daipom** [comment]
> If the old td-agent service is running, this pop-up appears. (Maybe it's been that way before. I'll check it.)
> 
> 2023-06-27_13h33_46

I confirmed it's been that way before. (This message appears though we don't need a reboot)
We should ignore this for now.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:42]: Somehow, this doesn't seem to work.
I confirmed the package behavior, but I couldn't find the old log files.

I'm reconsidering log file migration on Windows.
On Windows, even if the log file is moved, Fluentd will output the log to new files.
So, we may not need to move the log file.
(It remains in the backup-ed directory `td-agent.old`.)

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:57]: Somehow, this doesn't seem to work.
I'm investigating it now.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:53]:
```suggestion
  ) else (
```

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:50]: This condition looks reversed.
`File Found` => `!ERRORLEVEL! == 0` => `preserve`

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:57]: I have commented on the cause below.
However, if we should keep the old log files in `td-agent.old` without moving them, we should always create this directory. (commented above)

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat]: Thanks, I've fixed.

Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat]: Removing logic was omitted.

**Comment by kenhys** [comment]
Try to reduce a risk of losing files.

**Comment by kenhys** [comment]
changing path separator issue may be solved with "v=!v:\=/!" expression.

**Comment by kenhys** [comment]
Need to fix:

```
WixQuietExec64:  ボリューム シリアル番号は 7830-FC66 です
WixQuietExec64:  C:\OPT\TD-AGENT
WixQuietExec64:  └─etc
WixQuietExec64:      └─td-agent
WixQuietExec64:  "Create symlink c:\opt\fluent\var\log\td-agent to c:\opt\fluent\var\log\fluent"
WixQuietExec64:  指定されたパスが見つかりません。
WixQuietExec64:  "Create symlink c:\opt\td-agent\etc\td-agent to c:\opt\fluent\etc\fluent"
WixQuietExec64:  既に存在するファイルを作成することはできません。
WixQuietExec64:  Error 0x80070001: Command line returned an error.
WixQuietExec64:  Error 0x80070001: QuietExec64 Failed
WixQuietExec64:  Error 0x80070001: Failed in ExecCommon method
```

**Comment by kenhys** [comment]
msi

**Comment by daipom** [comment]
Thanks! I will review this again!

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:11]: I am wondering if it is possible for the service to be active at this time.
Against my expectations, it is judged to be active in my tests.
And `net stop` often fails.

```
WixQuietExec64:  "Active fluentdwinsvc service is detected"
WixQuietExec64:  The service is starting or stopping.  Please try again later.
```

Sometimes the installation does not work and this may have something to do with it.
I will check it.

Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: I found the cause. 
At `InstallFluentdWinSvc` action, the service is started because of the option `--reg-winsvc-auto-start`, which is added here.
I think this makes the behavior unstable.

https://github.com/fluent/fluent-package-builder/blob/f285807b93b03c8338dd0b58371c247e6fbfd045/fluent-package/msi/source.wxs#L179-L181

We should disable this auto-start since few users want this auto-start.

* No one would be happy to start Fluentd with the default settings.
* It opens some ports by default. It is undesirable from a security point of view.

How about adding fix like this?: https://github.com/daipom/fluent-package-builder/commit/10ae6750565df7b1f70e0098404c8336565ced0a

I'm checking the CI results on my fork with this fix now.

Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: The CI on my fork succeeds!: https://github.com/daipom/fluent-package-builder/actions/runs/5504485369

Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: I confirmed the package built in the CI works as expected in my environment.

Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: It looks better to me, then could you push a commit on this pull request?

Comment (ashie) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: In addition, please add description about removing `--reg-winsvc-auto-start` to the first comment.

Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: @kenhys Sure!

@ashie To the first comment of this PR?

Comment (ashie) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: > To the first comment of this PR?

Yes, it's used as the commit message on squash merging.

Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:11]: > > To the first comment of this PR?
> 
> Yes, it's used as the commit message on squash merging.

Instead of adding the change to this PR, I have made another PR:

* #521

I want to rebase this branch on it after it is merged.

Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat]: > It looks better to me, then could you push a commit on this pull request?

I have applied the changes except #521!

**Comment by daipom** [comment]
I have rebased this to apply

* #521

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [reply fluent-package/msi/assets/fluent-package-post-migration.bat]: Solved by #521.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:43]: Is this for debugging?
If we need file lists for debugging, then should we use `/F` option?

File: fluent-package/msi/assets/fluentd.bat
Comment (daipom) [inline fluent-package/msi/assets/fluentd.bat:11]: Somehow, this does not appear to work for

* https://github.com/fluent/fluentd/issues/1138

I tested the package with the following config, but backslash was still used and could not load the existing buffer files on restart.

Maybe we should handle this separately from this PR.

```xml
<source>
  @type sample
  tag test
</source>

<match test.**>
  @type stdout
  <buffer>
    @type file
    path "#{ENV['FLUENT_PACKAGE_TOPDIR']}/var/log/fluentd/buffer/stdout"
    flush_at_shutdown false
    flush_interval 10
  </buffer>
</match>
```

log

```
[info]: using configuration file: <ROOT>
  <source>
    @type sample
    tag "test"
  </source>
  <match test.**>
    @type stdout
    <buffer>
      @type "file"
      path "C:\\opt\\fluent\\/var/log/fluentd/buffer/stdout"
      flush_at_shutdown false
      flush_interval 10
    </buffer>
  </match>
</ROOT>
```

**Review by daipom — COMMENTED** [review]
Now the migration process works stably!

I have several concerns on a point other than the migration process itself.
I have commented them below.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (daipom) [inline fluent-package/msi/assets/fluent-package-post-migration.bat:12]: I checked this behavior by rewriting the codes.
When exiting ERRORLEVEL `1`, Rollback doesn't occur, and the installation is completed without running the migration process in this script.

I think this is acceptable for now.
Fixes to ensure Rollback should be done in future work.

Comment (ashie) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:12]: I think so too.

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:43]: Yes, it is for debugging use. it is good idea to apply /F

File: fluent-package/msi/assets/fluent-package-post-migration.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-post-migration.bat:12]: Thanks for the confirmation. :+1:

File: fluent-package/msi/assets/fluentd.bat
Comment (kenhys) [reply fluent-package/msi/assets/fluentd.bat]: it will be fixed via ab9df59

Comment (kenhys) [reply fluent-package/msi/assets/fluentd.bat]: Here is the example to verify scope of enabledelayedexpansion:

```
set FLUENT_PACKAGE_TOPDIR=%~dp0..\
set TD_AGENT_TOPDIR=%~dp0..\td-agent
setlocal enabledelayedexpansion
echo "FLUENT_PACKAGE_TOPDIR:"
set FLUENT_PACKAGE_TOPDIR=!FLUENT_PACKAGE_TOPDIR:\=/!
set TD_AGENT_TOPDIR=!TD_AGENT_TOPDIR:\=/!
echo %TD_AGENT_TOPDIR%
echo %FLUENT_PACKAGE_TOPDIR%
endlocal

echo "OUTSIDE"
echo %TD_AGENT_TOPDIR%
echo %FLUENT_PACKAGE_TOPDIR%
```

expansion

Comment (daipom) [reply fluent-package/msi/assets/fluentd.bat]: Wow! I see! Thanks for the fix!

Comment (daipom) [reply fluent-package/msi/assets/fluentd.bat]: I will check it after the CI finished and the new package is built.

**Comment by ashie** [comment]
Install test seems failed since ab9df591684fc63283da50df695f05f5ff0ac373
```
Installing C:\fluentd\fluent-package\msi\repositories\fluent-package-5.0.0-x64.msi 
Start-Service : Cannot find any service with service name 'fluentdwinsvc'. 
At C:\fluentd\fluent-package\msi\install-test.ps1:7 char:1
+ Start-Service fluentdwinsvc
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (fluentdwinsvc:String) [Start-Se  
   rvice], ServiceCommandException
    + FullyQualifiedErrorId : NoServiceFoundForGivenName,Microsoft.PowerShell.  
   Commands.StartServiceCommand
```

**Comment by kenhys** [comment]
Redundant expression causes registration error.
It will be fixed via 40bc79d.

**Comment by daipom** [comment]
Thanks! I will check the behavior of the package after CI finishes.

File: fluent-package/msi/assets/fluentd.bat
Comment (daipom) [reply fluent-package/msi/assets/fluentd.bat]: https://github.com/fluent/fluent-package-builder/pull/494#discussion_r1259005439 seems still not solved.
I checked the package, but the result is the same.

Sorry, there are some points I don't understand.

https://github.com/fluent/fluent-package-builder/pull/494#discussion_r1259179257 means the slash conversion is lost outside the scope?
If so, the environmental variable `ENV['FLUENT_PACKAGE_TOPDIR']` in the config should still contain backslashes?

I checked the previous environmental variable `ENV['TD_AGENT_TOPDIR']` in the config.
It returns `C:\\opt\\td-agent\\`.
I think it is strange...
Now it seems to me that no one sets this old variable anymore.
Why can Fluentd still access this variable...

I think it is correct that this old environment variable still works as is.
If not, then the old config using this variable can't work in v5.
But I don't understand how this still works now...

Comment (daipom) [reply fluent-package/msi/assets/fluentd.bat]: Summary

* `ENV['TD_AGENT_TOPDIR']` should work in the new package, but it is uncertain in the current implementation.
* The logic converting BachSlash in `ENV['FLUENT_PACKAGE_TOPDIR']` works for the environmental variables, but it would not be used from the configuration.

We should handle these points in other PRs.

**Review by daipom — APPROVED** [review]
LGTM.

We should handle the points I commented on https://github.com/fluent/fluent-package-builder/pull/494#discussion_r1263217312 in other PRs.

**Pull request merged**
